packages/databyss-editor/lib/clipboardUtils.tsx

```diff
@@ -0,0 +1,445 @@
+        ranges: rangesForBlockAfter,
+      },
+      type: block.type,
+    },
+  }
+}
+
+// checks is state selection is collapsed
+export const isSelectionCollapsed = (selection: Selection): boolean => {
+  const { anchor, focus } = selection
+  return anchor.index === focus.index && anchor.offset === focus.offset
+}
+
+// return atomic or new id
+const getId = (type: BlockType, id: string): string =>
+  isAtomicInlineType(type) ? id : new ObjectId().toHexString()
+
+// takes blocks array and resets the id's for non atomic types
+export const resetIds = (fragment: Block[]): Block[] =>
+  fragment.map(block => ({ ...block, _id: getId(block.type, block._id) }))
+
+const addBlockData = (frag: Block[]): Block[] =>
+  frag.map(b => ({ ...b, __showNewBlockMenu: false, __isActive: false }))
+
+// always have the anchor come before the focus
+export const sortSelection = (selection: Selection): Selection => {
+  const { anchor, focus } = selection
+  let _anchor = anchor
+  let _focus = focus
+  // always put anchor before focus
+  if (anchor.index > focus.index) {
+    _focus = [_anchor, (_anchor = _focus)][0]
+  } else if (anchor.offset > focus.offset && anchor.index === focus.index) {
+    _focus = [_anchor, (_anchor = _focus)][0]
+  }
+
+  return {
+    anchor: _anchor,
+    focus: _focus,
+  }
+}
+
+// returns fragment in state selection
+export const getCurrentSelection = (state: EditorState): Block[] => {
+  if (isSelectionCollapsed(state.selection)) {
+    return []
+  }
+
+  let frag: Block[] = []
+
+  const { blocks, selection } = state
+
+  const { anchor, focus } = sortSelection(selection)
+
+  const _blocks = cloneDeep(blocks)
+
+  // if selection is within the same block
+  if (anchor.index === focus.index) {
+    const _selectionLength = focus.offset - anchor.offset
+    const _block = blocks[anchor.index]
+    // split block at anchor offset and use `after`
+    const _firstSplit = splitBlockAtOffset({
+      block: _block,
+      offset: anchor.offset,
+    }).after
+
+    // split block at length of selection and get `before`
+    const _secondSplit = splitBlockAtOffset({
+      block: _firstSplit || _block,
+      offset: _selectionLength,
+    }).before
+
+    // if selection is use the first split
+    const _frag = _secondSplit || _firstSplit
+
+    if (_frag) {
+      frag.push({ ..._frag, _id: getId(_frag.type, blocks[anchor.index]._id) })
+    }
+  }
+
+  // if selection is more than one block
+  if (anchor.index < focus.index) {
+    // first block
+    const { after: firstBlock } = isAtomicInlineType(blocks[anchor.index].type)
+      ? { after: blocks[anchor.index] }
+      : splitBlockAtOffset({
+          block: blocks[anchor.index],
+          offset: anchor.offset,
+        })
+
+    if (firstBlock) {
+      frag.push({
+        ...firstBlock,
+        _id: getId(firstBlock.type, blocks[anchor.index]._id),
+      })
+    }
+
+    const _sliceLength = focus.index - anchor.index
+
+    if (_sliceLength > 1) {
```

Comment:
This is confusing. To handle all the blocks covered by the selection, can't we just use a forEach loop like:
```
blocks.forEach((block, index) => {
  if (index === anchor.index) {
    // do firstBlock stuff
  } else if (index === focus.index) {
    // do lastBlock stuff
  } else if (index > anchor.index && index < focus.index) {
    // do in between block stuff
  }
}
```